File: Blood on the Clocktower V3/engine/roles.py
"""
官方暗流涌动(Trouble Brewing)角色定义
基于钟楼百科 https://clocktower-wiki.gstonegames.com
"""

BOTC_ROLES = {
    "洗衣妇": {
        "team": "townsfolk",
        "ability": "游戏开始时会得知两名玩家中有一名特定镇民身份。",
        "first_night": True,
        "other_nights": False,
        "icon": "🧺",
    },
    "图书管理员": {
        "team": "townsfolk",
        "ability": "游戏开始时会得知两名玩家中有一名特定外来者身份，或得知本局没有外来者。",
        "first_night": True,
        "other_nights": False,
        "icon": "📚",
    },
    "调查员": {
        "team": "townsfolk",
        "ability": "游戏开始时会得知两名玩家中有一名特定爪牙身份。",
        "first_night": True,
        "other_nights": False,
        "icon": "🔍",
    },
    "厨师": {
        "team": "townsfolk",
        "ability": "游戏开始时会得知有多少对相邻的玩家是邪恶阵营。",
        "first_night": True,
        "other_nights": False,
        "icon": "👨‍🍳",
    },
    "共情者": {
        "team": "townsfolk",
        "ability": "每晚会得知与你相邻的存活玩家中有多少人是邪恶的(0-2)。",
        "first_night": True,
        "other_nights": True,
        "icon": "💕",
    },
    "占卜师": {
        "team": "townsfolk",
        "ability": "每晚会选择两名玩家，得知其中是否有恶魔(可选择一名已选过的玩家)。注意有干扰项机制。",
        "first_night": True,
        "other_nights": True,
        "icon": "🔮",
    },
    "送葬者": {
        "team": "townsfolk",
        "ability": "每晚会得知今天白天被处决的玩家的角色。",
        "first_night": False,
        "other_nights": True,
        "icon": "⚰️",
    },
    "僧侣": {
        "team": "townsfolk",
        "ability": "每晚会选择一名玩家(不能选自己)，该玩家当晚不会被恶魔杀害。",
        "first_night": False,
        "other_nights": True,
        "icon": "🙏",
    },
    "守鸦人": {
        "team": "townsfolk",
        "ability": "如果你在夜晚死亡，你可以选择一名玩家并得知其角色。",
        "first_night": False,
        "other_nights": False,
        "icon": "🐦",
    },
    "贞洁者": {
        "team": "townsfolk",
        "ability": "第一次被提名时，如果提名者是镇民则其立即被处决。",
        "first_night": False,
        "other_nights": False,
        "icon": "💎",
    },
    "猎手": {
        "team": "townsfolk",
        "ability": "白天可公开选择一名玩家，如果该玩家是恶魔则其立即死亡。能力只能用一次。",
        "first_night": False,
        "other_nights": False,
        "icon": "🏹",
    },
    "士兵": {
        "team": "townsfolk",
        "ability": "恶魔无法杀害你。",
        "first_night": False,
        "other_nights": False,
        "icon": "🛡️",
    },
    "镇长": {
        "team": "townsfolk",
        "ability": "如果白天只有3名存活玩家且无人被处决，你的阵营获胜。当恶魔攻击你时可能有一名其他玩家死亡。",
        "first_night": False,
        "other_nights": False,
        "icon": "🏛️",
    },
    "管家": {
        "team": "outsider",
        "ability": "每晚会选择一名玩家成为你的主人，你只能在主人投票后才能投票。",
        "first_night": True,
        "other_nights": True,
        "icon": "🕴️",
    },
    "酒鬼": {
        "team": "outsider",
        "ability": "你以为自己是某镇民角色，但实际不是，你得到的是错误信息。",
        "first_night": False,
        "other_nights": False,
        "icon": "🍺",
    },
    "陌客": {
        "team": "outsider",
        "ability": "你可能被判定为邪恶阵营或爪牙，即使你是善良的。",
        "first_night": False,
        "other_nights": False,
        "icon": "👤",
    },
    "圣徒": {
        "team": "outsider",
        "ability": "如果你被处决，你的阵营落败。",
        "first_night": False,
        "other_nights": False,
        "icon": "😇",
    },
    "投毒者": {
        "team": "minion",
        "ability": "每晚会选择一名玩家，该玩家今晚和明天白天中毒(获得错误信息)。",
        "first_night": True,
        "other_nights": True,
        "icon": "☠️",
    },
    "间谍": {
        "team": "minion",
        "ability": "你可以查看魔典。你可能被判定为善良阵营。",
        "first_night": True,
        "other_nights": True,
        "icon": "🕵️",
    },
    "红唇女郎": {
        "team": "minion",
        "ability": "如果恶魔死亡，你变成恶魔。",
        "first_night": False,
        "other_nights": False,
        "icon": "💋",
    },
    "男爵": {
        "team": "minion",
        "ability": "本局游戏额外增加2名外来者(减少2名镇民)。",
        "first_night": False,
        "other_nights": False,
        "icon": "👑",
    },
    "小恶魔": {
        "team": "demon",
        "ability": "每晚会选择一名玩家杀害(首夜不杀人)。如果自杀，一名爪牙变成恶魔。",
        "first_night": False,
        "other_nights": True,
        "icon": "😈",
    },
}

BOTC_TEAMS = {
    "townsfolk": ["洗衣妇", "图书管理员", "调查员", "厨师", "共情者", "占卜师", "送葬者", "僧侣", "守鸦人", "贞洁者", "猎手", "士兵", "镇长"],
    "outsider": ["管家", "酒鬼", "陌客", "圣徒"],
    "minion": ["投毒者", "间谍", "红唇女郎", "男爵"],
    "demon": ["小恶魔"],
}

BOTC_WIN_CONDITIONS = {
    "good": "所有恶魔均已死亡。",
    "evil": "场上只剩两名存活玩家(旅行者不计入)。",
}

# 官方夜晚顺序表(严格按wiki顺序)
# 首夜：间谍在洗衣妇之前查看魔典，这样间谍可以提前知道全角色信息来伪装
NIGHT_ORDER_FIRST = [
    ("黄昏", "检查所有玩家闭眼"),
    ("爪牙信息", "唤醒爪牙,指认恶魔(仅7人以上)"),
    ("恶魔信息", "唤醒恶魔,指认爪牙,展示3个不在场善良角色(仅7人以上)"),
    ("投毒者", "选择一名玩家下毒"),
    ("间谍", "查看魔典(知晓全场角色以制定伪装策略)"),
    ("洗衣妇", "得知两名玩家中一名特定镇民身份"),
    ("图书管理员", "得知两名玩家中一名特定外来者身份或告知无外来者"),
    ("调查员", "得知两名玩家中一名特定爪牙身份"),
    ("厨师", "得知相邻邪恶对数"),
    ("共情者", "得知相邻存活邪恶玩家数"),
    ("占卜师", "选择两名玩家,得知其中是否有恶魔(含干扰项)"),
    ("管家", "选择一名玩家为主人"),
    ("黎明", "所有玩家睁眼"),
]

NIGHT_ORDER_OTHER = [
    ("黄昏", "检查所有玩家闭眼"),
    ("投毒者", "选择一名玩家下毒"),
    ("僧侣", "选择一名玩家保护(不能被恶魔杀害)"),
    ("间谍", "查看魔典(知晓全场角色以制定伪装策略)"),
    ("红唇女郎", "若恶魔死亡则变成新恶魔(存活>=5人时)"),
    ("小恶魔", "选择一名玩家杀害(可自杀传位)"),
    ("守鸦人", "若今晚死亡则查看一名玩家的角色"),
    ("送葬者", "得知今天白天被处决的玩家角色"),
    ("共情者", "得知相邻存活邪恶玩家数"),
    ("占卜师", "选择两名玩家,得知其中是否有恶魔(含干扰项)"),
    ("管家", "选择一名玩家为主人"),
    ("黎明", "睁眼并宣布死亡情况"),
]

File: Blood on the Clocktower V5/botc_server/internal/engine/roles.go
package engine

// 阵营常量
const (
	TeamTownsfolk = "townsfolk"
	TeamOutsider  = "outsider"
	TeamMinion    = "minion"
	TeamDemon     = "demon"
)

// RoleDef 角色定义
type RoleDef struct {
	Key         string
	Name        string
	Team        string // 空 = 说书人伪步骤
	Ability     string
	FirstNight  bool
	OtherNights bool
	Icon        string
	// NeedsChoice 夜晚需要玩家主动选择目标：
	// 真人 → 挂起相位 + 前端选择 UI + 超时兜底；AI → 内部直通决策
	// 当前落地：占卜师；预留：僧侣/管家/投毒者/小恶魔
	NeedsChoice bool
}

// 夜晚行动顺序（严格遵循官方钟楼百科：暗流涌动）
// 首夜：无僧侣、无小恶魔（小恶魔首夜不行动）
var NightOrderFirst = []string{
	"minion_info", "demon_info", "poisoner", "spy",
	"washerwoman", "librarian", "investigator", "chef",
	"empath", "fortune_teller", "butler",
}

// 其他夜：红唇女郎/间谍为说书人提醒步骤（被动空走）
var NightOrderOther = []string{
	"poisoner", "monk", "spy", "scarlet_woman", "imp",
	"ravenkeeper", "undertaker", "empath", "fortune_teller", "butler",
}

// Roles 全部角色表（含说书人信息步骤）
var Roles = map[string]RoleDef{
	// ——— 说书人信息步骤 (伪角色) ————————————————
	"minion_info": {Key: "minion_info", Name: "爪牙信息", Ability: "爪牙得知恶魔", FirstNight: true, Icon: "📋"},
	"demon_info":  {Key: "demon_info", Name: "恶魔信息", Ability: "恶魔得知爪牙与三个不在场镇民", FirstNight: true, Icon: "📋"},

	// ——— 镇民 (13) ————————————————
	"washerwoman": {Key: "washerwoman", Name: "洗衣妇", Team: TeamTownsfolk,
		Ability:    "在你的首个夜晚，你会得知两名玩家，其中一名是某镇民。",
		FirstNight: true, Icon: "🧺"},
	"librarian": {Key: "librarian", Name: "图书管理员", Team: TeamTownsfolk,
		Ability:    "在你的首个夜晚，你会得知两名玩家，其中一名是某外来者（或没有外来者在场）。",
		FirstNight: true, Icon: "📚"},
	"investigator": {Key: "investigator", Name: "调查员", Team: TeamTownsfolk,
		Ability:    "在你的首个夜晚，你会得知两名玩家，其中一名是爪牙。",
		FirstNight: true, Icon: "🔍"},
	"chef": {Key: "chef", Name: "厨师", Team: TeamTownsfolk,
		Ability:    "在你的首个夜晚，你会得知有多少对相邻的存活玩家是邪恶阵营。",
		FirstNight: true, Icon: "👨‍🍳"},
	"empath": {Key: "empath", Name: "共情者", Team: TeamTownsfolk,
		Ability:    "每个夜晚，你会得知与你的两名相邻存活玩家中，有几人是邪恶阵营。",
		FirstNight: true, OtherNights: true, Icon: "💗"},
	"fortune_teller": {Key: "fortune_teller", Name: "占卜师", Team: TeamTownsfolk,
		Ability:    "每个夜晚，你要选择两名玩家（其中一人可以是自己）：你会得知其中是否有恶魔。",
		FirstNight: true, OtherNights: true, Icon: "🔮", NeedsChoice: true},
	"undertaker": {Key: "undertaker", Name: "送葬者", Team: TeamTownsfolk,
		Ability:     "每个夜晚*，你会得知今天因处决而死去的玩家的角色。",
		OtherNights: true, Icon: "⚰️"},
	"monk": {Key: "monk", Name: "僧侣", Team: TeamTownsfolk,
		Ability:     "每个夜晚*，你要选择一名玩家（与上个夜晚不同）：他当晚不会因恶魔而死。",
		OtherNights: true, Icon: "🧘", NeedsChoice: true},
	"ravenkeeper": {Key: "ravenkeeper", Name: "守鸦人", Team: TeamTownsfolk,
		Ability:     "如果你在夜晚死去，你醒来后会选择一名玩家，得知他的角色。",
		OtherNights: true, Icon: "🐦‍⬛"},
	"virgin": {Key: "virgin", Name: "贞洁者", Team: TeamTownsfolk,
		Ability: "当你首次被一名镇民提名时，该镇民立即被处决。",
		Icon:    "🌸"},
	"slayer": {Key: "slayer", Name: "猎手", Team: TeamTownsfolk,
		Ability: "每局游戏限一次，在白天，你可以公开选择一名玩家：如果他是恶魔，他死亡。",
		Icon:    "🏹"},
	"soldier": {Key: "soldier", Name: "士兵", Team: TeamTownsfolk,
		Ability: "你不会因恶魔而死。",
		Icon:    "🪖"},
	"mayor": {Key: "mayor", Name: "镇长", Team: TeamTownsfolk,
		Ability: "如果只有三名玩家存活，且当天没有玩家被处决，你的阵营获胜。",
		Icon:    "🏛️"},

	// ——— 外来者 (4) ————————————————
	"butler": {Key: "butler", Name: "管家", Team: TeamOutsider,
		Ability:    "每个夜晚，你要选择你的主人。只有主人投票时，你才能投票。",
		FirstNight: true, OtherNights: true, Icon: "🫖", NeedsChoice: true},
	"drunk": {Key: "drunk", Name: "酒鬼", Team: TeamOutsider,
		Ability:    "你以为你是一个镇民，但其实你不是。",
		FirstNight: true, OtherNights: true, Icon: "🍺"},
	"recluse": {Key: "recluse", Name: "陌客", Team: TeamOutsider,
		Ability: "你可能被当作邪恶阵营或恶魔。",
		Icon:    "🕳️"},
	"saint": {Key: "saint", Name: "圣徒", Team: TeamOutsider,
		Ability: "如果你被处决，善良阵营落败。",
		Icon:    "😇"},

	// ——— 爪牙 (4) ————————————————
	"poisoner": {Key: "poisoner", Name: "投毒者", Team: TeamMinion,
		Ability:    "每个夜晚，你要选择一名玩家：他的能力失效，并可能得到错误信息，直到下个黄昏。",
		FirstNight: true, OtherNights: true, Icon: "🧪", NeedsChoice: true},
	"spy": {Key: "spy", Name: "间谍", Team: TeamMinion,
		Ability:    "每个夜晚，你会查看魔典，得知所有玩家的角色。",
		FirstNight: true, OtherNights: true, Icon: "🕶️"},
	"scarlet_woman": {Key: "scarlet_woman", Name: "红唇女郎", Team: TeamMinion,
		Ability: "如果恶魔死亡且存活玩家不少于五人，你成为恶魔。",
		Icon:    "💃"},
	"baron": {Key: "baron", Name: "男爵", Team: TeamMinion,
		Ability: "场上会多两名外来者。",
		Icon:    "🎩"},

	// ——— 恶魔 (1) ————————————————
	"imp": {Key: "imp", Name: "小恶魔", Team: TeamDemon,
		Ability:     "每个夜晚*，你要选择一名玩家：他死亡。如果你因任何原因死亡，且场上有爪牙存活，一名爪牙会成为小恶魔。",
		OtherNights: true, Icon: "😈", NeedsChoice: true},
}

// GetRolesByTeam 按阵营获取角色
func GetRolesByTeam(team string) []string {
	out := []string{}
	for k, v := range Roles {
		if v.Team == team {
			out = append(out, k)
		}
	}
	return out
}
